File: standard-udfs-examples/standard-udfs-example-udfs-spark/src/test/scala/com/linkedin/stdudfs/examples/spark/TestStructCreateByNameFunctionWrapper.scala
/**
 * BSD 2-CLAUSE LICENSE
 *
 * Copyright 2018 LinkedIn Corporation.
 * All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package com.linkedin.stdudfs.examples.spark

import com.linkedin.stdudfs.spark.common.AssertSparkExpression._
import org.testng.annotations.{BeforeClass, Test}

class TestStructCreateByNameFunctionWrapper {

  @BeforeClass
  def registerFunction(): Unit = {
    registerStandardUdf(
      "struct_create_by_name",
      classOf[StructCreateByNameFunctionWrapper]
    )
  }

  @Test
  def testStructCreateByNameFunction(): Unit = {
    assertFunction("cast(struct_create_by_name('a', 'x', 'b', 'y') as struct<a:string, b:string>).a", "x")
    assertFunction("cast(struct_create_by_name('a', 'x', 'b', 'y') as struct<a:string, b:string>).b", "y")
    assertFunction("struct_create_by_name(null, 'x', 'b', 'y')", null)
    assertFunction("struct_create_by_name('a', 'x', null, 'y')", null)
  }
}

File: transportable-udfs-examples/transportable-udfs-example-udfs-avro/src/test/java/com/linkedin/transport/examples/avro/TestStructCreateByNameFunctionWrapper.java
/**
 * Copyright 2018 LinkedIn Corporation. All rights reserved.
 * Licensed under the BSD-2 Clause license.
 * See LICENSE in the project root for license information.
 */
package com.linkedin.transport.examples.avro;

import com.google.common.collect.ImmutableList;
import org.apache.avro.Schema;
import org.testng.annotations.Test;

import static com.linkedin.transport.avro.common.AssertAvroUdf.*;
import static com.linkedin.transport.avro.common.SchemaFactory.*;


public class TestStructCreateByNameFunctionWrapper {
  @Test
  public void testStructCreateByName() {
    assertFunction(new StructCreateByNameFunctionWrapper(), new Schema[]{STRING, STRING, STRING, STRING},
        new Object[]{"a", "x", "b", "y"}, ImmutableList.of("x", "y"));

    assertFunction(new StructCreateByNameFunctionWrapper(), new Schema[]{STRING, STRING, STRING, STRING},
        new Object[]{null, "x", "b", "y"}, null);

    assertFunction(new StructCreateByNameFunctionWrapper(), new Schema[]{STRING, STRING, STRING, STRING},
        new Object[]{"a", null, "b", "y"}, null);

    assertFunction(new StructCreateByNameFunctionWrapper(), new Schema[]{STRING, STRING, STRING, STRING},
        new Object[]{"a", "x", null, "y"}, null);

    assertFunction(new StructCreateByNameFunctionWrapper(), new Schema[]{STRING, STRING, STRING, STRING},
        new Object[]{"a", "x", "b", null}, null);
  }
}
